src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java

```diff
@@ -0,0 +1,451 @@
+    ConsumerCommittedResponse response = new ConsumerCommittedResponse();
+    response.offsets = new Vector<TopicPartitionOffsetMetadata>();
+    lock.writeLock().lock();
+    try {
+      if (consumer != null) {
+        for (io.confluent.kafkarest.entities.TopicPartition t : request.partitions) {
+          TopicPartition partition = new TopicPartition(t.getTopic(), t.getPartition());
+          OffsetAndMetadata offsetMetadata = consumer.committed(partition);
+          if (offsetMetadata != null) {
+            response.offsets.add(
+                new TopicPartitionOffsetMetadata(partition.topic(), partition.partition(),
+                    offsetMetadata.offset(), offsetMetadata.metadata()));
+          }
+        }
+      }
+    } finally {
+      lock.writeLock().unlock();
+    }
+    return response;
+  }
+
+
+  public boolean expired(long nowMs) {
+    return expiration <= nowMs;
+  }
+
+  public void updateExpiration() {
+    this.expiration = config.getTime().milliseconds() +
+        config.getInt(KafkaRestConfig.CONSUMER_INSTANCE_TIMEOUT_MS_CONFIG);
+  }
+
+  public long untilExpiration(long nowMs) {
+    return this.expiration - nowMs;
+  }
+
+  public KafkaRestConfig getConfig() {
+    return config;
+  }
+
+  public void setConfig(KafkaRestConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public int compareTo(KafkaConsumerState o) {
+    if (this.expiration < o.expiration) {
+      return -1;
+    } else if (this.expiration == o.expiration) {
+      return 0;
+    } else {
+      return 1;
+    }
+  }
+
+  /**
+   * An iterator / cursor to track the current position of the records sent to the client so far
+   */
+  public Iterator<ConsumerRecord<ClientK, ClientV>> getIterator() {
+    if (consumerRecords == null) {
+      return null;
+    }
+    if (iter == null) {
+      iter = consumerRecords.iterator();
+    }
+    return iter;
+  }
+
+  /**
+   * Initiate poll(timeout) request to retrieve consummer records, or return the existing
+   * consumer records if the records have not been fully consumed by client yet
+   */
+  public void getOrCreateConsumerRecords(long timeout) {
+    try {
+      lock.writeLock().lock();
+      if (!hasNext()) {
+        //reset index
+        this.index = 0;
+        consumerRecordList = new ArrayList<ConsumerRecord<ClientK, ClientV>>();
+        consumerRecords = consumer.poll(timeout);
+        //drain the iterator and buffer to list
+        for (ConsumerRecord<ClientK, ClientV> consumerRecord : consumerRecords) {
+          consumerRecordList.add(consumerRecord);
+        }
+      }
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  public ConsumerRecord<ClientK, ClientV> peek() {
+    if (hasNext()) {
+      ConsumerRecord<ClientK, ClientV> record = consumerRecordList.get(this.index);
+      return record;
+    }
+    return null;
+  }
+
+  public boolean hasNext() {
+    return consumerRecordList != null && this.index < consumerRecordList.size();
```

Comment:
I don't see any reason why a Java queue can't be used instead of list and building peek(), hasNext() and next() functions on top a list